dipy/align/vector_fields.pyx

```diff
@@ -3065,3 +3065,411 @@ def create_sphere(cnp.npy_intp nslices, cnp.npy_intp nrows,
                 else:
                     s[k,i,j] = 0
     return s
+
+
+cdef void _gradient_3d(floating[:,:,:] img, double[:,:] img_affine_inv,
+                       double[:] img_spacing, double[:,:] domain_affine,
+                       floating[:,:,:,:] out, int[:,:,:] inside):
+    r""" Gradient of a 3D image in physical space coordinates
+
+    Each grid cell (i, j, k) in the sampling domain (determined by
+    out.shape) is mapped to its corresponding physical point (x, y, z) by
+    multiplying domain_affine (its grid-to-space transform) by (i, j, k), then
+    the image is interpolated, at
+
+    P1=(x + h, y, z), Q1=(x - h, y, z)
+    P2=(x, y + h, z), Q2=(x, y - h, z)
+    P3=(x, y, z + h), Q3=(x, y, z - h)
+
+    (by mapping Pi and Qi to the grid using img_affine_inv: the inverse of the
+    grid-to-space transform of img). The displacement parameter h is of
+    magnitude 0.5 (in physical space units), therefore the approximated partial
+    derivatives are given by the difference between the image interpolated at
+    Pi and Qi.
+
+    Parameters
+    ----------
+    img : array, shape (S, R, C)
+        the input volume whose gradient will be computed
+    img_affine_inv : array, shape (4, 4)
+        the space-to-grid transform matrix associated to img
+    img_spacing : array, shape (3,)
+        the spacing between voxels (voxel size along each axis) of the input
+        volume
+    domain_affine : array, shape (4, 4)
```

Comment:
I agree, can you find better terms here?

If I understand correctly, there is an implied grid (out shape), a vox2mm (grid-to-space) transform for that grid (currently domain_affine).  Then there is an image, `img`, with an mm2vox (space-to-grid) transform.  You are going to fill the out shape grid by finding corresponding voxel (grid) points in `img` and finding the gradient by sampling at those points.

Maybe `out_grid2mm` for `domain_affine`?  `mm2img_grid` for `img_affine_inv`?

I guess you can get the voxel sizes `img_spacing` from `img_affine_inv`.